content/blog/rlang-1-0-0-errors/index.Rmd

```diff
@@ -0,0 +1,179 @@
+
+1. Fully committing to the display of errors as bullet lists
+1. Including the erroring function call by default, as in base R
+1. Embracing chained errors to represent contextual information
+
+If you'd like to try the new error style on your computer, install the development versions of rlang and dplyr (the latter needs to be adapted to the new error style) from github with:
+
+```{r, eval = FALSE}
+pak::pkg_install(c("r-lib/rlang", "tidyverse/dplyr"))
+```
+```{r setup, message = FALSE}
+library(rlang)
+library(dplyr)
+```
+
+
+## Displaying errors as bullet lists
+
+`rlang::abort()` makes it easy to structure error messages as a __bullet list__. We believe that errors should be both informative about the context of the error and easy to skim. A bullet list arrangement that lays out important pieces of information line by line provides the right trade off for this:
+
+```{r, error = TRUE}
+abort(c(
+  "This is the error message.",
+  "*" = "This is a bullet.",
+  "*" = "This is another bullet."
+))
+```
+
+The bullet symbol can be customised to provide a clue about the kind of information contained in the bullet. Use "ℹ" bullets to provide contextual information or hints, and "✖" bullets to state a problematic input or state.
+
+```{r, error = TRUE}
+abort(c(
+  "This is the error message.",
+  "x" = "Can't do this.",
+  "i" = "You could do that instead."
+))
+```
+
+Here is a dplyr example of an informative error message structured as a bullet list:
+
+```{r, error = TRUE}
+mtcars %>%
+  group_by(cyl) %>%
+  mutate(new = rep(am, 2))
+```
+
+In all bullet lists, the main error message (the error header in rlang terms) is indicated with a "!" sign. This makes it easy to skim for error headers in a stuffy R output.
+
+
+## Displaying the erroring function
+
+By default, `base::stop()` shows the function in which it was called:
+
+```{r, error = TRUE}
+add1 <- function(x) {
+  if (!is.numeric(x)) {
+    stop("`x` must be numeric.")
+  }
+  x + 1
+}
+
+add1("foo")
+```
+
+In rlang, we initially decided to turn off that feature because quite often the erroring function is unrelated to the function called by the user. This happens for instance when `stop()` or `abort()` are called from a helper function:
+
+```{r, error = TRUE}
+add1 <- function(x) {
+  check_numeric(x)
+  x + 1
+}
+
+check_numeric <- function(x) {
+  if (!is.numeric(x)) {
+    stop("`x` must be numeric.")
+  }
+}
+
+add1("foo")
+```
+
+To avoid distracting users with irrelevant information, `abort()` just didn't include a call in the error. However, we were missing out on contextual information that could help users understand the origin of an error without having to look at backtrace.
```

Comment:
```suggestion
To avoid distracting users with irrelevant information, `abort()` just didn't include a call in the error. However, we were missing out on contextual information that could help users understand the origin of an error without having to look at the back trace, and that context is particularly important in a long pipeline of function calls.`
```